### ISSUE TITLE
LoRA training loop via candle-lora

### ISSUE THREAD

**Issue by boersmamarcel** [boersmamarcel/chatty2#166]
## Context

The core adaptation mechanism. LoRA (Low-Rank Adaptation) fine-tunes only ~2M parameters (rank-8 on a 3B model) while keeping the 3B base model frozen. This makes training feasible as a background task on consumer hardware — even on CPU.

From the [s3 system in Jiang et al. 2025](https://doi.org/10.48550/arXiv.2512.16301): competitive performance with only **2,400 training examples** and modest compute.

## Dependencies

```toml
candle-lora = { git = "https://github.com/EricLBuehler/candle-lora.git", optional = true }
candle-lora-transformers = { git = "https://github.com/EricLBuehler/candle-lora.git", optional = true }
```

### Training Config

```rust
pub struct LoraTrainingConfig {
    pub lora_rank: usize,              // Default: 8
    pub lora_alpha: f64,               // Default: 16.0
    pub lora_dropout: f64,             // Default: 0.05
    pub learning_rate: f64,            // Default: 2e-4
    pub num_epochs: usize,             // Default: 3
    pub batch_size: usize,             // Default: 4
    pub warmup_steps: usize,           // Default: 10
    pub max_grad_norm: f64,            // Default: 1.0
    pub training_type: TrainingType,
}

pub enum TrainingType {
    Sft,                               // Supervised fine-tuning
    Dpo { beta: f64 },                 // Direct preference optimization (beta default: 0.1)
}
```

### Training Loop (SFT)

```rust
pub struct LoraTrainer {
    config: LoraTrainingConfig,
    device: candle_core::Device,
}

impl LoraTrainer {
    /// Run SFT training. Returns path to saved adapter.
    pub async fn train_sft(
        &self,
        base_model: &CandleLocalModel,
        dataset: &SftDataset,
        progress: impl Fn(TrainingProgress),
    ) -> Result<PathBuf>;

    /// Run DPO training. Returns path to saved adapter.
    pub async fn train_dpo(
        &self,
        base_model: &CandleLocalModel,
        dataset: &DpoDataset,
        progress: impl Fn(TrainingProgress),
    ) -> Result<PathBuf>;
}

pub struct TrainingProgress {
    pub epoch: usize,
    pub step: usize,
    pub total_steps: usize,
    pub loss: f64,
    pub learning_rate: f64,
    pub elapsed_secs: f64,
}
```

### SFT Loss

Standard cross-entropy on non-masked labels:
```
L_SFT = -Σ log P(y_i | x, y_<i)  for all i where label != -100
```

### DPO Loss

Direct preference optimization (no reward model needed):
```
L_DPO = -log(σ(β * (log π(chosen|x) - log π_ref(chosen|x) - log π(rejected|x) + log π_ref(rejected|x))))
```

- `π_ref` = base model (frozen, forward pass only)
- `π` = base model + LoRA (trainable)
- Single forward pass on both completions, log-prob arithmetic, sigmoid

### Training Trigger

Automatic threshold-based triggering:

```rust
pub struct TrainingScheduler {
    threshold: usize,  // e.g., 50 new examples
}

impl TrainingScheduler {
    /// Check if enough new data has accumulated since last training run
    pub fn should_train(&self, trace_store: &TraceStore) -> bool;

    /// Schedule a background training run
    pub fn schedule_training(
        &self,
        trainer: &LoraTrainer,
        model: &CandleLocalModel,
        trace_store: &TraceStore,
    ) -> Task<Result<PathBuf>>;
}
```

### Adapter Storage

```
~/.../chatty/models/adapters/
    ├── phi3_sft_v001.safetensors
    ├── phi3_sft_v002.safetensors
    ├── phi3_dpo_v001.safetensors
    └── training_log.json              # History of training runs
```

### Background Execution

- Training runs on a dedicated low-priority Tokio blocking task
- Cancellable via `Arc<AtomicBool>` (same pattern as StreamManager)
- Progress reported via events for potential future UI
- Does NOT block inference — base model can still serve requests during training

### Training Metrics

Logged to `training_log.json` and TraceStore's `training_runs` table:
- Loss per step and per epoch
- Validation loss (on held-out split)
- Training duration
- Number of examples used
- Adapter file path and size

## Acceptance Criteria

- [ ] `LoraTrainer` with SFT training loop
- [ ] `LoraTrainer` with DPO training loop
- [ ] LoRA adapters saved as `.safetensors` with versioned naming
- [ ] Training runs on background thread (doesn't block UI or inference)
- [ ] Cancellation support via `AtomicBool` flag
- [ ] Progress reporting (loss, step, epoch)
- [ ] `TrainingScheduler` with threshold-based auto-triggering
- [ ] Training metrics logged to `training_runs` table
- [ ] Gradient clipping and learning rate warmup
- [ ] Works with Phi-3, Gemma-2, Qwen2.5 architectures
- [ ] Validates on held-out split after each epoch
- [ ] Memory-efficient: only LoRA params in optimizer state

## Depends On
- #163 (CandleLocalModel — base model for training)
- #165 (DatasetBuilder — training data)

## Blocks
- #155 Issue 12 (Adapter management & hot-swap)
